src/shims/posix/sync.rs

```diff
@@ -112,15 +112,23 @@ fn mutex_get_or_create_id<'mir, 'tcx: 'mir>(
     ecx: &mut MiriEvalContext<'mir, 'tcx>,
     mutex_op: &OpTy<'tcx, Tag>,
 ) -> InterpResult<'tcx, MutexId> {
-    let id = mutex_get_id(ecx, mutex_op)?.to_u32()?;
-    if id == 0 {
-        // 0 is a default value and also not a valid mutex id. Need to allocate
-        // a new mutex.
+    let value_place = ecx.offset_and_layout_to_place(mutex_op, 4, ecx.machine.layouts.u32)?;
+    let (old, success) = ecx
+        .atomic_compare_exchange_scalar(
+            &value_place,
+            &ImmTy::from_uint(0u32, ecx.machine.layouts.u32),
+            ecx.mutex_next_id().to_u32_scalar().into(),
+            AtomicRwOp::Relaxed,
+            AtomicReadOp::Relaxed,
+            false,
+        )?
+        .to_scalar_pair()?;
+
+    if success.to_bool().expect("compare_exchange's second return value is a bool") {
         let id = ecx.mutex_create();
```

Comment:
What I don't like about this is that it is abstraction-breaking: we rely on `mutex_create` to use the same ID as what `mutex_next_id` returned above. Currently this is true since Miri is entirely sequential, but... ugh this doesn't make me very comfortable.

*At least* we should have an `assert!` here to check that the `id` we are returning is the one we just set. Better probably would be to make `mutex_crate` *take* an ID. For now it can just panic if the ID it gets isn't the one it would have sued anyway, but at least this makes for a more reasonable API overall.

Reply:
I refactored this to use `*_get_or_create()` functions that take in a closure which does the RMW